### PR TITLE
L2 bridge Implementation

### PR DESCRIPTION
**Pull request by SchSeba**
This PR Change the bridge plugin to allow empty IPAM.

When the IPAM dictionary is empty the bridge plugin will connect a pure L2 interface(without any IP address on the interface) from the pod to a bridge on the host.

This PR will allow containers and everything running on top of them to have a pure L2 interface connected to a bridge. This can be useful for use cases of sniffing, running dhclient, give the pod a trunk interface, etc..

### CODE REVIEW

File: plugins/main/bridge/README.md [hunk 1]
```diff
@@ -28,6 +28,16 @@ If the bridge is missing, the plugin will create one on first use and, if gatewa
 }
 ```
 
+## Example for l2 configuration
```
Comment (dankenigsberg) [inline plugins/main/bridge/README.md]: I'd call this "L2-only configuration"

File: plugins/main/bridge/README.md [hunk 2]
```diff
@@ -39,5 +49,5 @@ If the bridge is missing, the plugin will create one on first use and, if gatewa
 * `ipMasq` (boolean, optional): set up IP Masquerade on the host for traffic originating from this network and destined outside of it. Defaults to false.
 * `mtu` (integer, optional): explicitly set MTU to the specified value. Defaults to the value chosen by the kernel.
 * `hairpinMode` (boolean, optional): set hairpin mode for interfaces on the bridge. Defaults to false.
-* `ipam` (dictionary, required): IPAM configuration to be used for this network.
+* `ipam` (dictionary, required): IPAM configuration to be used for this network(for pure l2 network create empty dictionary).
```
Comment (dankenigsberg) [inline plugins/main/bridge/README.md]: Nicer English, imo:
```
 IPAM configuration to be used for this network. For L2-only network, create empty dictionary.
```

File: plugins/main/bridge/bridge.go [hunk 3]
```diff
@@ -358,103 +360,109 @@ func cmdAdd(args *skel.CmdArgs) error {
 		return err
 	}
 
-	// run the IPAM plugin and get back the config to apply
-	r, err := ipam.ExecAdd(n.IPAM.Type, args.StdinData)
-	if err != nil {
-		return err
-	}
+	// Assume l2 interface only
```
Comment (dankenigsberg) [inline plugins/main/bridge/bridge.go]: Please use uppercase L2 for consistency and legibility.

Comment (SchSeba) [reply plugins/main/bridge/bridge.go]: Done

File: plugins/main/bridge/README.md (hunk 2 above)
Comment (SchSeba) [reply plugins/main/bridge/README.md]: Done

File: plugins/main/bridge/README.md (hunk 1 above)
Comment (SchSeba) [reply plugins/main/bridge/README.md]: Done

**Comment by dcbw** [comment]
@SchSeba thanks for the update!  Can you add some testcases though?  Let's make sure we don't break anything.

File: plugins/main/bridge/bridge.go [hunk 4]
```diff
@@ -485,41 +493,44 @@ func cmdDel(args *skel.CmdArgs) error {
...
-	if err != nil {
-		return err
-	}
+		// There is a netns so try to clean up. Delete can be called multiple times
+		// so don't return an error if the device is already removed.
+		// If the device isn't there then don't try to clean up IP masq either.
+		var ipnets []*net.IPNet
+		err = ns.WithNetNSPath(args.Netns, func(_ ns.NetNS) error {
+			var err error
+			ipnets, err = ip.DelLinkByNameAddr(args.IfName)
```
Comment (squeed) [inline plugins/main/bridge/bridge.go]: It looks like the container interface is only deleted in Layer3 mode? That doesn't seem right.

File: plugins/main/bridge/bridge.go [hunk 5]
```diff
@@ -485,41 +493,44 @@ func cmdDel(args *skel.CmdArgs) error {
...
-	var ipnets []*net.IPNet
-	err = ns.WithNetNSPath(args.Netns, func(_ ns.NetNS) error {
-		var err error
-		ipnets, err = ip.DelLinkByNameAddr(args.IfName)
-		if err != nil && err == ip.ErrLinkNotFound {
-			return nil
+	isLayer3 := n.IPAM.Type != ""
+
+	if isLayer3 {
+		if err := ipam.ExecDel(n.IPAM.Type, args.StdinData); err != nil {
```
Comment (bboreham) [inline plugins/main/bridge/bridge.go]: This line needs to stay before `if args.Netns == "" {` which is on line 496 as of my review.

Reason is that we want to clean up ipam even when the namespace has disappeared.

File: plugins/main/bridge/bridge.go [hunk 6]
```diff
@@ -358,103 +360,109 @@ func cmdAdd(args *skel.CmdArgs) error {
...
+				ipam.ExecDel(n.IPAM.Type, args.StdinData)
+				os.Setenv("CNI_COMMAND", "ADD")
+			}
+		}()
 
-	// Configure the container hardware address and IP address(es)
-	if err := netns.Do(func(_ ns.NetNS) error {
-		contVeth, err := net.InterfaceByName(args.IfName)
+		// Convert whatever the IPAM result was into the current Result type
+		result, err = current.NewResultFromResult(r)
```
Comment (squeed) [inline plugins/main/bridge/bridge.go]: You're setting result.interfaces twice. This variable should be renamed to `ipamResult`.

Comment (SchSeba) [reply plugins/main/bridge/bridge.go]: I first assume the interface is L2 so I create a L2 result 

```go
// Assume L2 interface only
result := &current.Result{CNIVersion: cniVersion}
result.Interfaces = []*current.Interface{brInterface, hostInterface, containerInterface}
```

But if the IPAM is configured then 

```go
// Convert whatever the IPAM result was into the current Result type
result, err = current.NewResultFromResult(r)
```

So I can use one variable only for the return 

```go
return types.PrintResult(result, cniVersion)
```

And don't need to check again if the interface is a L2 to decide what variable to return 

```go
if  isLayer3 {
 return types.PrintResult(ipamResult, cniVersion)
}

return types.PrintResult(result, cniVersion)
```

File: plugins/main/bridge/bridge.go (hunk 4 above)
Comment (SchSeba) [reply plugins/main/bridge/bridge.go]: Done

File: plugins/main/bridge/bridge.go (hunk 6 above)
Comment (bboreham) [reply plugins/main/bridge/bridge.go]: @squeed this makes sense - do you still want it changed?

Comment (squeed) [reply plugins/main/bridge/bridge.go]: I guess I'm still a bit uncomfortable with the structure of the code. The fact that the `result` defined in line 364 is *maybe* overwritten strikes me as a source of later bugs. In other words, someone else might manipulate `result`, not understanding that their changes are maybe going to be overwritten.

Could you only define `result` once and update it as needed?

**Comment by squeed** [comment]
@SchSeba thanks! that's exactly what I was thinking about.

One last thing: Could you add a test case, please?

**Comment by squeed** [comment]
Also, you'll need to rebase on master to fix a testcase flake.

**Comment by dcbw** [comment]
@SchSeba could you rebase to fix the test flake and update testcases per @squeed's comment?  Thanks!

**Comment by SchSeba** [comment]
@squeed Can you please check if the tests are good or you want me to check something else?

File: plugins/main/bridge/README.md [hunk 7]
```diff
@@ -28,6 +28,17 @@ If the bridge is missing, the plugin will create one on first use and, if gatewa
 }
 ```
 
+## L2-only configuration
```
Comment (matthewdupre) [inline plugins/main/bridge/README.md]: I think "Example L2-only configuration" would be a better title?

**Comment by squeed** [comment]
Looks good to me. Can you squash down to a single commit?

File: plugins/main/bridge/README.md (hunk 7 above)
Comment (SchSeba) [reply plugins/main/bridge/README.md]: Done

**Comment by SchSeba** [comment]
@squeed now it looks better or you want me to remove also the merge commits?

**Comment by dcbw** [comment]
@SchSeba yes, please rebase against master and squash to remove the merge commits.  In general, instead of merging master into your topic branch, please rebase your topic branch on top of master.

**Comment by SchSeba** [comment]
@dcbw  Done.

**Pull request merged**